include/arch/riscv/arch/32/mode/hardware.h

```diff
@@ -69,40 +69,38 @@
 
 static inline uint64_t riscv_read_time(void)
 {
-    uint32_t nH1, nL, nH2;
+    uint32_t nH_prev, nH, nL;
     asm volatile(
         "rdtimeh %0\n"
         "rdtime  %1\n"
         "rdtimeh %2\n"
-        : "=r"(nH1), "=r"(nL), "=r"(nH2));
-    if (nH1 < nH2) {
+        : "=r"(nH_prev), "=r"(nL), "=r"(nH));
+    if (nH != nH_prev) {
         /* Ensure that the time is correct if there is a rollover in the
          * high bits between reading the low and high bits. */
         asm volatile(
             "rdtime  %0\n"
             : "=r"(nL));
-        nH1 = nH2;
     }
-    return ((uint64_t)((uint64_t) nH1 << 32)) | (nL);
+    return ((uint64_t)nH << 32) | nL;
 }
 
 static inline uint64_t riscv_read_cycle(void)
 {
-    uint32_t nH1, nL, nH2;
+    uint32_t nH_prev, nH, nL;
     asm volatile(
         "rdcycleh %0\n"
         "rdcycle  %1\n"
         "rdcycleh %2\n"
-        : "=r"(nH1), "=r"(nL), "=r"(nH2));
-    if (nH1 != nH2) {
+        : "=r"(nH_prev), "=r"(nL), "=r"(nH));
+    if (nH != nH_prev) {
         /* Ensure that the time is correct if there is a rollover in the
          * high bits between reading the low and high bits. */
         asm volatile(
             "rdcycle  %0\n"
```

Comment:
Is there actually a strong reason to read the low value again and not just use `0` here? Rationale would be that there is some jitter on the low value anyway, but 0 would be closer to the actual point in time where we wanted to read things in the first place.

Reply:
It seems like a reasonable optimization when you know that the value of `rdcycle`must have contained 0 at some time between the two reads of `rdcycleh`.